MTM-49965: reference-notifications: rework of non-persistent subscriptions and tokens (#720)
SoftwareAG/c8y-docs: content/reference/notifications-bundle/overview.md

and, if that is set to the value `true`, the created token will be shared.
If this body parameter is not present or `false`, a token will be exclusive (not shared) by default.

If a consumer's token is *not* shared, the consumer is an **exclusive** consumer.
Only one consumer client can connect using an exclusive token: attempts to connect further consumers with the same exclusive token will result in error.
An exclusive consumer will receive a copy of all notifications from the subscription its token is for.

If a consumer's token is shared, the consumer is a **shared** consumer. Additional consumer clients can connect using the same token.
When only one shared consumer is connected, it receives a copy of all notifications from subscription.
As additional consumer clients connect using the same token, the consumers' notification load is rebalanced so that 
each consumer receives a non-overlapping subset (share) of the notifications from the subscription. 
The collective ensemble of consumers sharing a token can be thought of as one logical consumer.
Collectively, the ensemble will receive all notifications for the subscription. 

> The notification load is divided into shares dictated by a hash of the id of the source that generated the notification (typically a device id).

Reviewer: Not sure I would mention hashing in user documentation
```suggestion
The notification load is spread across the shared consumers according to the id of the source that generated the notification, typically a device id.
All notifications for a given id will be delivered to the same consumer, and each consumer may be receiving notifications for many different ids.
```

Reviewer 2: Agree, writing about hash may just bring us more questions from customers.

This means there is no benefit using shared tokens unless the notifications feeding the subscription are coming from multiple sources.

> Note that, as the shares are divided by a hash of the ids, it can result in an asymmetric balance of notification load across the shares if there are few source ids involved. This should even out for higher numbers of source ids.

Reviewer: Again I just don't want to say "hash" here.  Not certain that "algorithm" is any better although people probably have a sense of what that means even if they couldn't define it.
```suggestion
Note that the load spreading algorithm may result in an asymmetric balance of notification load across the shares when there are few source ids in the subscription.
Load should generally become more evenly distributed as the number of sources increases.
```

> In order to help keep the messages from a given set of source ids sticky to consumer client shares in the face of consumer client connection interruptions, the consumer clients can provide an 

Reviewer: ```suggestion
In order to help keep the messages from a given set of source ids "sticky" to shared consumers in the face of connection interruptions, the consumer clients can provide an 
```

optional `consumer` parameter in their connection URL string, in addition to their usual `token` parameter. 
For example: two consumers identifying themselves as *instance1* and *instance2* would connect using URL paths
`notification2/consumer?token=xyz&consumer=instance1` and `notification2/consumer?token=xyz&consumer=instance2`.

> Subscriptions are always unaware of the nature and number of their consumers: any number of shared and exclusive tokens can created for the same subscription and they all operate independently, each receiving their own copy of the notifications. This means you can have multiple shared tokens for the same subscription and their load will only be divided up within the scope of each shared token.

Reviewer 2: `tokens can created` -> `tokens can be created`?

Author: Will change ...

Author: Done

They use replicated secondary storage to maintain large backlogs (within the constraints of any configured backlog limits)
and to maintain the consumers' positions in subscription notification streams.
When a consumer of a persistent subscription has their connection interrupted,
whether that is due to network issues or deliberate actions by the consumer,
upon reconnection they will continue to receive notifications from the position they were at before the outage 
(specifically, from the message after the last one they acknowledged successfully before the outage). 

Non-persistent subscriptions are only buffered in memory and their consumers' positions are not persisted across disconnections of the consumer.
When a consumer of a non-persistent subscription has their connection interrupted,
upon reconnection they will start receiving notifications from the most recent message of the subscription,
missing all other notifications that occurred during the connection outage. 
This will be the case for such temporarily disconnected consumers,
even if other consumers of the same non-persistent subscription 
are still receiving older messages that occurred while it was not connected. 

> When both a persistent and a non-persistent subscription are created with the same name, that is, with the same `subscription` body parameter value in the request, 

Reviewer 3: ```suggestion
If you create both a persistent and a non-persistent subscription with the same name, that is, with the same `subscription` body parameter value in the request, 
```

> they will be *separate*, independent subscriptions. Such subscriptions can vary by any other body parameters they choose: they do not *have* to be

Reviewer 3: ```suggestion
they are *separate*, independent subscriptions. Such subscriptions can vary by any other body parameters you choose. They do not have to be
```

> persistent and non-persistent variations of the same notification data, although to avoid confusion it is recommended to keep such subscriptions identical except for the `nonPersistent` parameter.  

Reviewer 3: ```suggestion
persistent and non-persistent variations of the same notification data. However, we recommend you to keep such subscriptions identical except for the `nonPersistent` parameter to avoid confusion.  
```

When a consumer creates a token for either a persistent or non-persistent subscription, 

> it **must** distinguish which type of subscription it is targeting by use of the `non-persistent` body parameter in the token creation request.

Reviewer 3: ```suggestion
it must distinguish which type of subscription it is targeting by using the `non-persistent` body parameter in the token creation request.
```

> Like for the subscription, this body parameter defaults to `false`, making the token target a persistent subscription. 

Reviewer 3: ```suggestion
For the subscription, this body parameter defaults to `false`, making the token target a persistent subscription. 
```

Setting the body parameter to `true` in the token creation request targets a non-persistent subscription.  

> ### Shared tokens

Reviewer 3: ```suggestion
### Shared tokens

```

Shared tokens allow parallelization of the consumer client workload for a notification subscription.
This is useful if the notifications would otherwise arrive at a higher rate than the consuming client application can process them.

> It has no impact on the rate of notification throughput within, and thus their egress from, Cumulocity core.

Reviewer 3: ```suggestion
It has no impact on the rate of notification throughput within, and thus their egress from, {{< product-c8y-iot >}} core.
```

> When creating a token, an optional Boolean `shared` body parameter can be added to the request,

Reviewer 3: ```suggestion
When you create a token, you can add the optional Boolean body parameter `shared` to the request.
```

> and, if that is set to the value `true`, the created token will be shared.

Reviewer 3: ```suggestion
If it is set to `true`, the created token is shared.
```

> If this body parameter is not present or `false`, a token will be exclusive (not shared) by default.

Reviewer 3: ```suggestion
If it is not present or `false`, the token is exclusive (not shared).
```

> If a consumer's token is *not* shared, the consumer is an **exclusive** consumer.

Reviewer 3: ```suggestion
If a consumer's token is not shared, the consumer is an *exclusive* consumer.
```

> Only one consumer client can connect using an exclusive token: attempts to connect further consumers with the same exclusive token will result in error.

Reviewer 3: ```suggestion
Only one consumer client can connect using an exclusive token. An attempt to connect further consumers with the same exclusive token results in an error.
```

> An exclusive consumer will receive a copy of all notifications from the subscription its token is for.

Reviewer 3: ```suggestion
An exclusive consumer receives a copy of all notifications from the subscription its token is for.
```

> If a consumer's token is shared, the consumer is a **shared** consumer. Additional consumer clients can connect using the same token.

Reviewer 3: ```suggestion
If a consumer's token is shared, the consumer is a *shared consumer*. Additional consumer clients can connect using the same token.
```

> When only one shared consumer is connected, it receives a copy of all notifications from subscription.

Reviewer 3: ```suggestion
If only one shared consumer is connected, it receives a copy of all notifications from the subscription.
```

As additional consumer clients connect using the same token, the consumers' notification load is rebalanced so that 
each consumer receives a non-overlapping subset (share) of the notifications from the subscription. 

> The collective ensemble of consumers sharing a token can be thought of as one logical consumer.

Reviewer 3: ```suggestion
The set of consumers sharing a token can be thought of as a single logical consumer.
```

> Collectively, the ensemble will receive all notifications for the subscription. 

Reviewer 3: ```suggestion
Collectively, the set receives all notifications for the subscription. 
```

> The notification load is spread across the shared consumers according to the id of the source that generated the notification, typically a device id.

Reviewer 3: ```suggestion
The notification load is spread across the shared consumers according to the ID of the source that generated the notification, typically a device ID.
```

> All notifications for a given id will be delivered to the same consumer, and each consumer may be receiving notifications for many different ids.

Reviewer 3: ```suggestion
All notifications for a given ID will be delivered to the same consumer. Each consumer may receive notifications for many different IDs.
```

> This means there is no benefit using shared tokens unless the notifications feeding the subscription are coming from multiple sources.

Reviewer 3: ```suggestion
This means that there is no benefit using shared tokens unless the notifications feeding the subscription are coming from multiple sources.
```

> Note that the load spreading algorithm may result in an asymmetric balance of notification load across the shares when there are few source ids in the subscription.

Reviewer 3: ```suggestion
Note that the load spreading algorithm may result in an asymmetric balance of notification load across the shares when there are few source IDs in the subscription.
```

> Load should generally become more evenly distributed as the number of sources increases.

Reviewer 3: ```suggestion
The load should generally become more evenly distributed as the number of sources increases.
```

> In order to help keep the messages from a given set of source ids "sticky" to shared consumers in the face of connection interruptions, the consumer clients can provide an 

Reviewer 3: ```suggestion
In order to help keep the messages from a given set of source IDs stick to shared consumers in the face of connection interruptions, the consumer clients can provide an 
```

optional `consumer` parameter in their connection URL string, in addition to their usual `token` parameter. 

> For example: two consumers identifying themselves as *instance1* and *instance2* would connect using URL paths

Reviewer 3: ```suggestion

For example: two consumers identifying themselves as *instance1* and *instance2* connect using URL paths
```

`notification2/consumer?token=xyz&consumer=instance1` and `notification2/consumer?token=xyz&consumer=instance2`.

> Subscriptions are always unaware of the nature and number of their consumers: any number of shared and exclusive tokens can be created for the same subscription and they all operate independently, each receiving their own copy of the notifications. This means you can have multiple shared tokens for the same subscription and their load will only be divided up within the scope of each shared token.

Reviewer 3: ```suggestion
Subscriptions are always unaware of the nature and number of their consumers: any number of shared and exclusive tokens can be created for the same subscription and they all operate independently, each receiving their own copy of the notifications. This means you can have multiple shared tokens for the same subscription and their load is only divided within the scope of each shared token.
```

This way, information like the subscription name can be extracted and the create token REST point can be called again, all on the client side.
The {{< product-c8y-iot >}} microservice Java SDK [TokenApi](https://github.com/SoftwareAG/cumulocity-clients-java/blob/develop/java-client/src/main/java/com/cumulocity/sdk/client/messaging/notifications/TokenApi.java) class contains a public refresh method which is implemented purely on the client side.

### Non-persistent subscriptions and their tokens

> When creating a subscription, an optional Boolean `nonPersistent` field can be added to the request body, 

Reviewer: Is "field" the right name for this entity?  According to https://swagger.io/docs/specification/2-0/describing-request-body/ "body parameter" might be more correct, although not sure what term we use elsewhere in the documentation.

Author: I'll remove 'body' form after request on this change (otherwise 2 bodies). So will be '... body parameter can be added to the request, ...'

Author: .. probably same for shared so will amend that and anywhere else I've changed also

Author: done

and, if that field is set to the value `true`, the created subscription will be non-persistent. 
If this field is not present or `false`, a subscription will be persistent by default.

Persistent subscriptions ensure consumers never miss a message if their connection is interrupted.

> They use replicated secondary storage to maintain large backlogs (within the constraints of any maximum backlog sizes)

Reviewer: Personal wording preference, feel free to ignore.
```suggestion
They use replicated secondary storage to maintain large backlogs (within the constraints of any configured backlog limits)
```

and to maintain the consumers' positions in subscription notification streams.
When a consumer of a persistent subscription has their connection interrupted,
whether that is due to network issues or deliberate actions by the consumer,

> upon reconnection they will continue to receive notifications from position they were at before the outage 

Reviewer: ```suggestion
upon reconnection they will continue to receive notifications from the position they were at before the outage 
```

(specifically, from the message after the last one they acknowledged successfully before the outage). 

> Non-persistent subscriptions are only buffered in memory and their consumers' positions are not persisted (not durable).

Reviewer: ```suggestion
Non-persistent subscriptions are only buffered in memory and their consumers' positions are not persisted across disconnections of the consumer.
```

When a consumer of a non-persistent subscription has their connection interrupted,
upon reconnection they will start receiving notifications from the most recent message of the subscription,
missing all other notifications that occurred during the connection outage. 
This will be the case for such temporarily disconnected consumers,
even if other consumers of the same non-persistent subscription 
are still receiving older messages that occurred while it was not connected. 

> When both a persistent and a non-persistent subscriptions are created with the same name (`subscription` field value in the request body) 

Reviewer: Possibly too many commas here now but the parentheses didn't quite feel right...
```suggestion
When both a persistent and a non-persistent subscription are created with the same name, that is, with the same `subscription` field value in the request body, 
```

they will be *separate*, independent subscriptions. Such subscriptions can vary by any other fields they choose: they do *not* have to be
persistent and non-persistent variations of the same notification data, though using such similar subscriptions in notably different ways 

> may be surprising to others, so it may not be wise to do this unless they are otherwise the same. 

Reviewer: Agree with @slc-softwareag's comment elsewhere that this could be stated more concisely/clearly.

Author: Reworked (push shortly) and also the next paragraph, which was written in the context of subscriptions with the same name/subscription-field, but has been updated to apply to all subscriptions (you have to use right non-persistent toekn value to target a notification whether there is name commonality or not)

Author: pushed

When a consumer creates a token for either of these, 
it can distinguish which subscription it is targeting by use of the `non-persistent` field in the token creation request.
Like for the subscription, this field defaults to `false`, making the token target the persistent subscription. 

> Setting this field to `true` in the token creation request targets the non-persistent one.  

Reviewer: I think this could be said more simply and firmly.  Maybe something like this: "To create a token for a non-persistent subscription, it is *required* to include a `non-persistent` body parameter in the request, with its value set to `true`.  Otherwise, the request will be taken to refer to a persistent subscription with the same name."

Author: Ah, may have just done so (following on from comment above). Please review again - your alternative might be preferable

Author: done - but please recriticise as I had already taken initiative on this

### Shared tokens

> Shared tokens allow parallelization of the consumer workload. 

Reviewer: Not sure about this one but it felt like it needed a little more.
```suggestion
Shared tokens allow parallelization of the consumer workload for a notification subscription. 
```

Author: Done

This is useful if the notifications would otherwise arrive at a higher rate than the consumer can process them.

> It has no impact on the rate of notification throughput through, and thus their egress from, Cumulocity. 

Reviewer: ```suggestion
It has no impact on the rate of notification throughput through, and thus their egress from, the Cumulocity core.
```

Author: Done and have also already change 'throughput through' to 'throughput within' but not quite sure that is the best option (it's better).
